aten/src/ATen/native/TopKImpl.h

```diff
@@ -49,36 +49,60 @@ void topk_impl_loop(
       if (largest) {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
+            if ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first)) return true;
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return false;
+            return x.second < y.second;
           });
       } else {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first));
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return true;
```

Comment:
Note that when using `std::partial_sort` in the case of `auto use_partial_sort = k * 64 <= n`, it will be stable sorted by default. We can do stable sort without adding to the complexity since we already keep the idx of each element.

Reply:
Yes, that makes sense since you are checking both values and indices.